hci: add advertiser support for ManufacturerData and ServiceData

deadprogram opened the pull request:
This PR adds advertiser support for ManufacturerData and ServiceData to the HCI implementation.

Addresses #306

deadprogram committed:
hci: add advertiser support for ManufacturerData and ServiceData

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram committed:
gap/linux: correct HCI implementation so it can Configure/Start/Stop advertising correctly.
Needed to update ServiceData after Advertisment is already started.

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram committed:
examples: add example to broadcast advertising servicedata

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram commented on gap_linux.go:
```diff
@@ -125,6 +125,26 @@ func (a *Advertisement) Stop() error {
 }
 
+// SetServiceData sets the service data for the advertisement.
+func (a *Advertisement) SetServiceData(sd []ServiceDataElement) error {
```
@aykevl do you happen to know how I can do this? :smile_cat:

aykevl replied:
What is your goal?

If you want to change advertisement data then I think you can stop and restart the advertisement with the new service data. Looking at the API I don't see a straightforward way to update any part of the advertisement. But this is straightforward to do from userspace: simply stop and restart the advertisement with the new parameters.

See:
https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/doc/org.bluez.LEAdvertisement.rst
https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/doc/org.bluez.LEAdvertisingManager.rst

deadprogram replied:
> If you want to change advertisement data then I think you can stop and restart the advertisement 

OK I will probably do that in a separate PR.

deadprogram replied:
@aykevl I took your advice and got rid of `SetServiceData()` entirely. PTAL.

aykevl commented on gap_nrf51.go:
```diff
@@ -72,6 +72,12 @@ func (a *Advertisement) Stop() error {
 
+// SetServiceData sets the service data for the advertisement.
+func (a *Advertisement) SetServiceData(sd []ServiceDataElement) error {
+	// TODO: implement
+	return nil
```
As long as these are unimplemented, I'd say they need to return an error.

deadprogram replied:
Good point. Changed everywhere that it is not yet implemented.

deadprogram replied:
Oops I missed that one. Done.

aykevl commented on gap_nrf528xx-advertisement.go:
```diff
@@ -84,3 +84,9 @@ func (a *Advertisement) Stop() error {
+
+// SetServiceData sets the service data for the advertisement.
+func (a *Advertisement) SetServiceData(sd []ServiceDataElement) error {
+	// TODO: implement
+	return nil
```
Looks like this can be implemented here, but it needs a different API.
https://docs.nordicsemi.com/bundle/s132_v6.0.0_api/page/group_b_l_e_g_a_p_f_u_n_c_t_i_o_n_s.html#ga9969047f4e7485c3f856c841978cc31a
Instead of a `SetServiceData`, you'd need to construct an entirely new advertisement payload (`bluetooth.AdvertisementOptions`) and set it as a whole.

deadprogram replied:
How about preserve the initial options and then set a new payload when `SetServiceData` is changed?

Please see the new example `beacon` for how I am trying to use this function.

deadprogram replied:
See my other comment above.

colececil commented on gap_hci.go:
```diff
@@ -420,3 +441,51 @@ func (a *Advertisement) Start() error {
 	return a.adapter.hci.leSetAdvertiseEnable(false)
 }
+
+// SetServiceData sets the service data for the advertisement.
```
Here are some points I think it might be helpful to cover in the method documentation here, to make it more clear how `SetServiceData` needs to be used and what its effects are:

1. It causes a new advertisement to be broadcast, with the updated service data that has been passed in.
2. Just updating the data in the `[]ServiceDataElement` that has already been passed in to `Configure` via `AdvertisementOptions.ServiceData` will not change the data being broadcast in the advertisement. `SetServiceData` must be called for the new data to go into effect.
3. What happens if `Configure` or `Start` have not yet been called on the `Advertisement`?
4. What effect does this have if the `Advertisement` has been configured with the `Interval` option?

deadprogram commented:
After some consideration, I think that it is better to just call `adv.Stop()`, then `adv.Configure()` and then `adv.Start()` again instead of adding a new API function.

This makes it more straightforward to understand what is happening, and to consequently then make the various platform implementations work as expected.

@colececil I think this implementation answers the questions you raised:

> It causes a new advertisement to be broadcast, with the updated service data that has been passed in.

That is accommodated by the pattern in this updated PR. 

> Just updating the data in the []ServiceDataElement that has already been passed in to Configure via AdvertisementOptions.ServiceData will not change the data being broadcast in the advertisement. SetServiceData must be called for the new data to go into effect.

That is no longer a problem since there is only 1 way to set this data.

> What happens if Configure or Start have not yet been called on the Advertisement?

This is also not a problem now.

> What effect does this have if the Advertisement has been configured with the Interval option?

This is just a normal advertising option, nothing specific to the `ServiceData`.

I have updated this PR accordingly, including the new example. It works correctly with the `HCI`, I will still need to make some changes for other targets.

deadprogram committed:
gap/linux: update implementation for GAP to allow for calling adv.Configure()
followed by adv.Start() followed by adv.Stop() multiple times. This is required
in order to update advertising ServiceData.

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram commented:
I just added another commit to this PR that allows the `example/beacon` to run correctly on Linux.

deadprogram commented:
OK this PR is rebased and squashed. Works correctly on HCI and Linux. nrf is currently broken (See #310) and Windows does not support `ServiceData`. macOS does not yet have the `Advertisement` implemented. As such, I think this PR is now ready for final review.

deadprogram commented:
@colececil @aykevl or anyone else, any thoughts here? I think this is ready to merge.

colececil commented:
@deadprogram I haven't had the chance to check it out in detail, but the new approach you described does indeed sound more straightforward and less confusing.

A couple questions: 

1. To send updated data at a set interval, I'm assuming I would leave the `Interval` option unset as before, and just call call `Stop`, `Configure`, and `Start` at whatever interval I want?
2. If the above assumption is correct, is there any advantage to calling `Stop` right after the advertisement is sent, rather than right before sending the next one? Like, would that save power or anything? And if so, is there a way for the program to know when the advertisement has been sent, so it knows when `Stop` can safely be called? 

Thanks!

deadprogram commented:
> To send updated data at a set interval, I'm assuming I would leave the Interval option unset as before, and just call call Stop, Configure, and Start at whatever interval I want?

Not setting `Interval` is probably just leaving it at the default. So as long as it is something less than the frequency that you are calling "Stop, Configure, and Start" it would do what you would expect.

 > If the above assumption is correct, is there any advantage to calling Stop right after the advertisement is sent, rather than right before sending the next one?

That would almost certainly result in some receivers not getting the advertisement data, since you do not know how often they are scanning for updates..

deadprogram commented:
Rebased against the latest `dev` and ready.

deadprogram committed:
gap: make generic services function for HCI separate function

Signed-off-by: deadprogram <ron@hybridgroup.com>

deadprogram commented:
OK I realized from some testing I needed to add another commit. The problem was that multiple calls to `adv.Configure()` when using the HCI adaptor were creating additional generic Services each time it was called. The additional commit handles that case.

deadprogram committed:
fix: ensure that HCI advertising interval is set to a default value, otherwise different adverting types do not work correctly

Signed-off-by: deadprogram <ron@hybridgroup.com>

bgould commented on gap_linux.go:
```diff
@@ -45,8 +46,8 @@ func (a *Adapter) DefaultAdvertisement() *Advertisement {
-	if a.properties != nil {
-		panic("todo: configure advertisement a second time")
+	if a.started {
+		panic("cannot configure advertisement that is already started")
```
I wonder if this condition should return an error instead of panicking.

deadprogram replied:
That seems reasonable. I will make that change.

deadprogram replied:
Done. Thanks for the feedback!

bgould commented:
Overall this is working great for me on `arduino-nano33` to implement the "BTHome" protocol using this program - https://github.com/bgould/tinygo-float-sensor/blob/main/main.go ... Home Assistant is able to automatically discover the device and receive sensor values from the advertising data every 100ms.  Going to let it run for a while and log data for a bit, but so far it is working well.  Will also try this on pico-w and let you know the results.

Was also able to confirm using AdvertisementType option appears to working as expected using nRF Connect.  At some point it might be nice to add this option for SoftDevice as well.

Thank you @deadprogram, this is really good :)

deadprogram commented:
>  Will also try this on pico-w and let you know the results.

@bgould advertising is not working correctly on Pico-W at present. See https://github.com/tinygo-org/bluetooth/issues/301

deadprogram commented:
> At some point it might be nice to add this option for SoftDevice as well.

@bgould softdevice advertising is also currently non-functional. See https://github.com/tinygo-org/bluetooth/issues/310

deadprogram commented:
I've addressed all of the feedback, and squashed commits as much as makes sense. Now going to merge this so further work can proceed.

Thanks very much to @colececil @aykevl @bgould for feedback and testing!

deadprogram commented:
FYI for @colececil @bgould or anyone else interested in BTHome, I just started this little helper package https://github.com/hybridgroup/go-bthome

colececil commented:
That's awesome @deadprogram! Thanks!

bgould commented:
> > At some point it might be nice to add this option for SoftDevice as well.
> 
> @bgould softdevice advertising is also currently non-functional. See #310

FWIW ... advertisements using SoftDevice are working with this program on `feather-nrf52840` - https://github.com/bgould/tinygo-float-sensor/blob/main/main.go ... that said I am able to reproduce the crash described in #310 on the same board, but it does not appear to be directly related to advertisement data.

deadprogram commented:
@bgould that is interesting to know, might help to track down the problem there...

colececil commented:
@deadprogram Good news! I updated my project (https://github.com/colececil/automatic-soil-monitor) to use the revised API here, along with the BTHome helper package you created, and it's still working as expected! And it's working well with Home Assistant:

![image](https://github.com/user-attachments/assets/83ad1460-3c79-43c8-bd53-b25e90ba86a4)

deadprogram commented:
That is good to hear, thank you for testing @colececil !
